app/src/main/java/com/burhanrashid52/photoediting/FileSaveHelper.kt

```diff
@@ -0,0 +1,152 @@
+import android.net.Uri
+import android.os.Build
+import android.provider.MediaStore
+import androidx.appcompat.app.AppCompatActivity
+import androidx.lifecycle.*
+import java.io.IOException
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+
+/**
+ * General contract of this class is to
+ * create a file on a device.
+ *
+ * How to Use it-
+ * Call [FileSaveHelper.createFile]
+ * if file is created you would receive it's file path and Uri
+ * and after you are done with File call [FileSaveHelper.notifyThatFileIsNowPubliclyAvailable]
+ *
+ * Remember! in order to shutdown executor call [FileSaveHelper.addObserver] or
+ * create object with the [FileSaveHelper.FileSaveHelper]
+ */
+class FileSaveHelper(private val mContentResolver: ContentResolver) : LifecycleObserver {
+    private val executor: ExecutorService? = Executors.newSingleThreadExecutor()
+    private val fileCreatedResult: MutableLiveData<FileMeta> = MutableLiveData()
+    private var resultListener: OnFileCreateResult? = null
+    private val observer = Observer { fileMeta: FileMeta ->
+        if (resultListener != null) {
+            resultListener!!.onFileCreateResult(fileMeta.isCreated,
+                    fileMeta.filePath,
+                    fileMeta.error,
+                    fileMeta.uri)
+        }
+    }
+
+    constructor(activity: AppCompatActivity) : this(activity.contentResolver) {
+        addObserver(activity)
+    }
+
+    private fun addObserver(lifecycleOwner: LifecycleOwner) {
+        fileCreatedResult.observe(lifecycleOwner, observer)
+        lifecycleOwner.lifecycle.addObserver(this)
+    }
+
+    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
+    fun release() {
+        executor?.shutdownNow()
+    }
+
+    /**
+     * The effects of this method are
+     * 1- insert new Image File data in MediaStore.Images column
+     * 2- create File on Disk.
+     *
+     * @param fileNameToSave fileName
+     * @param listener       result listener
+     */
+    fun createFile(fileNameToSave: String, listener: OnFileCreateResult?) {
+        resultListener = listener
+        executor!!.submit {
+            val cursor: Cursor? = null
+            val filePath: String
+            try {
+                val newImageDetails = ContentValues()
+                val imageCollection = buildUriCollection(newImageDetails)
+                val editedImageUri = getEditedImageUri(fileNameToSave, newImageDetails, imageCollection)
+                filePath = getFilePath(cursor, editedImageUri)
+                updateResult(true, filePath, null, editedImageUri, newImageDetails)
+            } catch (ex: Exception) {
+                ex.printStackTrace()
+                updateResult(false, null, ex.message, null, null)
+            } finally {
+                cursor?.close()
+            }
+        }
+    }
+
+    private fun getFilePath(cursor: Cursor?, editedImageUri: Uri?): String {
+        var cursor = cursor
```

Comment:
So the original code reassigns `cursor`, which is probably not intended. The auto-converter attempted to correct here.

We can leave this for now but lets leave a TODO here so we can take a look at what this is doing later.

Reply:
Here the converter was forced to create a local version of the cursor variable, because the parameter incoming comes as VAL and those can't be reassigned as the code tries to do right after

Reply:
> Here the converter was forced to create a local version of the cursor variable, because the parameter incoming comes as VAL and those can't be reassigned as the code tries to do right after

I understand. My point is the reassignment in the original code is probably unintended, and we should address later.